src/core/model/table/dynamic_position_list_index.cpp

```diff
@@ -0,0 +1,176 @@
+            ++valid_records_number_;
+        }
+    }
+    valid_records_number_ -= deleted_records_ids.size();
+
+    for (auto it = clusters_.begin(); it != clusters_.end();) {
+        auto& [cluster_value, cluster] = *it;
+        ProcessDeleteToCluster(cluster, deleted_records_ids);
+        auto new_cluster_it = new_clusters.find(cluster_value);
+        if (new_cluster_it != new_clusters.end()) {
+            auto const& [new_cluster_value, new_cluster] = *new_cluster_it;
+            cluster.insert(cluster.end(), new_cluster.begin(), new_cluster.end());
+            new_clusters.erase(new_cluster_value);
+        }
+
+        if (cluster.empty()) {
+            it = clusters_.erase(it);
+        } else {
+            sort(cluster.begin(), cluster.end());
+            ++it;
+        }
+    }
+
+    for (auto& new_cluster : new_clusters) {
+        clusters_.emplace(std::move(new_cluster));
+    }
+}
+
+std::unordered_map<int, unsigned> DynamicPositionListIndex::CreateFrequencies(
+        Cluster const& cluster, std::vector<int> const& probing_table) {
+    std::unordered_map<int, unsigned> frequencies;
+
+    for (int index : cluster) {
+        frequencies[probing_table[index]]++;
+    }
+
+    return frequencies;
+}
+
+std::shared_ptr<std::vector<int> const> DynamicPositionListIndex::CalculateAndGetProbingTable()
+        const {
+    if (probing_table_cache_ != nullptr) return probing_table_cache_;
+    cluster_values_cache_.clear();
+    std::vector<int> probing_table = std::vector<int>(next_record_id_);
+    int next_cluster_id = 1;
+    cluster_values_cache_.push_back({});
+    for (auto& [value, cluster] : clusters_) {
+        int value_id = next_cluster_id++;
+        for (int position : cluster) {
+            probing_table[position] = value_id;
+        }
+        cluster_values_cache_.emplace_back(value);
+    }
+
+    return std::make_shared<std::vector<int>>(probing_table);
+}
+
+std::unique_ptr<DynamicPositionListIndex> DynamicPositionListIndex::Intersect(
+        DynamicPositionListIndex const* that) const {
+    assert(this->next_record_id_ == that->next_record_id_);
+
+    if (this->valid_records_number_ > that->valid_records_number_) {
+        return that->Probe(this);
+    } else {
+        return this->Probe(that);
+    }
+}
+
+std::unique_ptr<DynamicPositionListIndex> DynamicPositionListIndex::Probe(
+        DynamicPositionListIndex const* that) const {
+    std::shared_ptr<std::vector<int> const> probing_table = that->CalculateAndGetProbingTable();
+    std::vector<ClusterValue> const& cluster_values = that->GetClusterValuesCache();
+    assert(this->next_record_id_ == probing_table->size());
+    ClusterCollection new_clusters;
+    unsigned int new_size = 0;
+
+    std::unordered_map<int, std::pair<ClusterValue, Cluster>> partial_clusters;
+
+    for (auto& [value, cluster] : clusters_) {
+        for (int position : cluster) {
+            int probing_table_value_id = (*probing_table)[position];
```

Comment:
UB if the probing table is old and the record is one of the newly inserted?
That is:
1. The initial FD verification is done. `Intersect` was called, the probing table cache is now active.
2. A new record was inserted. Not accounted for in the probing table.
3. A follow-up check is performed. One of the clusters contains the new record, leading to UB here.

Maybe the caching is not necessary and we should unconditionally create and update the probing table along with the clusters?